File: app/src/main/java/com/thesunnahrevival/sunnahassistant/data/local/ReminderDAO.java
package com.thesunnahrevival.sunnahassistant.data.local;

import android.database.sqlite.SQLiteConstraintException;
import android.net.Uri;

import com.thesunnahrevival.sunnahassistant.data.model.AppSettings;
import com.thesunnahrevival.sunnahassistant.data.model.Reminder;

import java.util.List;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;


@Dao
public interface ReminderDAO {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertReminder(Reminder reminder);

    @Delete
    void deleteReminder(Reminder reminder);

    @Query("SELECT * FROM reminders_table WHERE timeInSeconds >= :offsetFromMidnight AND ((day == :day AND month == :month AND year == :year) OR (day == :day AND month == 12 AND year == 0) OR day == 0 OR customScheduleDays LIKE '%' || :nameOfTheDay || '%')  ORDER BY isEnabled DESC, timeInSeconds")
    LiveData<List<Reminder>> getUpcomingReminders(long offsetFromMidnight, String nameOfTheDay, int day, int month, int year);

    @Query("SELECT * FROM reminders_table WHERE timeInSeconds <= :offsetFromMidnight AND ((day == :day AND month == :month AND year == :year) OR (day == :day AND month == 12 AND year == 0) OR day == 0 OR customScheduleDays LIKE '%' || :nameOfTheDay || '%')  ORDER BY isEnabled DESC, timeInSeconds")
    LiveData<List<Reminder>> getPastReminders(long offsetFromMidnight, String nameOfTheDay, int day, int month, int year);

    @Query("SELECT * FROM reminders_table WHERE ((day == :day AND month == :month AND year == :year) OR (day == :day AND month == 12 AND year == 0) OR day == 0 OR customScheduleDays LIKE '%' || :nameOfTheDay || '%')  ORDER BY isEnabled DESC, timeInSeconds")
    LiveData<List<Reminder>> getRemindersOnDay(String nameOfTheDay, int day, int month, int year);

    @Query("SELECT * FROM reminders_table WHERE (category == 'Prayer' AND (day == :day AND month == :month AND year =:year)) ORDER BY timeInSeconds")
    LiveData<List<Reminder>> getPrayerTimes(int day, int month, int year);

    @Query("SELECT * FROM reminders_table WHERE frequency == 'Weekly'")
    LiveData<List<Reminder>> getWeeklyReminders();

    @Query("SELECT * FROM reminders_table WHERE frequency == 'Monthly'")
    LiveData<List<Reminder>> getMonthlyReminder();

    @Query("SELECT * FROM reminders_table WHERE frequency == 'One Time'")
    LiveData<List<Reminder>> getOneTimeReminders();

    @Query("SELECT * FROM reminders_table WHERE ((day == :day AND month == :month AND year == :year) OR (day == :day AND month == 12 AND year == 0) OR day == 0 OR customScheduleDays LIKE '%' || :nameOfTheDay || '%') AND isEnabled ORDER BY timeInSeconds")
    List<Reminder> getRemindersOnDayValue(String nameOfTheDay, int day, int month, int year);

    @Query("SELECT * FROM reminders_table WHERE (category == 'Prayer' AND (day == :day AND month == :month AND year =:year)) ORDER BY timeInSeconds")
    List<Reminder> getPrayerTimesValue(int day, int month, int year);

    @Query("UPDATE reminders_table SET isEnabled =:isEnabled WHERE id ==:id")
    void setEnabled(int id, boolean isEnabled);

    @Query("UPDATE reminders_table SET isEnabled =:isEnabled WHERE reminderName ==:prayerName")
    void setPrayerTimeEnabled(String prayerName, boolean isEnabled);

    @Query("UPDATE reminders_table SET `offset` =:offsetValue, reminderName =:newPrayerName, reminderInfo =:reminderInfo WHERE reminderName == :prayerName")
    void updatePrayerTimeDetails(String prayerName, String newPrayerName, String reminderInfo, int offsetValue);

    @Query("UPDATE reminders_table SET month =:month, year =:year, timeInSeconds =:timeInSeconds WHERE id == :id")
    void updateGeneratedPrayerTimes(int id, int month, int year, long timeInSeconds);

    @Insert
    void addRemindersList(List<Reminder> remindersList);

    default String addRemindersListIfNotExists(List<Reminder> remindersList) {
        try {
            addRemindersList(remindersList);
        }
        catch (SQLiteConstraintException e){
            return "Reminders Already Added";
        }
        return "Successfully Added";
    }
    @Query("DELETE FROM reminders_table WHERE category == 'Prayer' ")
    void deleteAllPrayerTimes();

    @Query("SELECT * FROM reminders_table WHERE timeInSeconds > :offsetFromMidnight AND ((day == :day AND month == :month AND year == :year) OR (day == :day AND month == 12 AND year == 0) OR day == 0 OR customScheduleDays LIKE '%' || :nameOfTheDay || '%') AND isEnabled ORDER BY isEnabled DESC, timeInSeconds")
    Reminder getNextScheduledReminder(long offsetFromMidnight, String nameOfTheDay, int day, int month, int year);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertSettings(AppSettings settings);

    @Query("SELECT * FROM app_settings")
    LiveData<AppSettings> getAppSettings();

    @Query("SELECT * FROM app_settings")
    AppSettings getAppSettingsValue();

    @Update
    void updateAppSettings(AppSettings appSettings);


    @Query("SELECT showNextReminderNotification FROM app_settings")
    boolean getIsForegroundEnabled();

    @Query("UPDATE reminders_table SET category =:newCategory WHERE category == :deletedCategory")
    void updateCategory(String deletedCategory, String newCategory);

    @Query("UPDATE app_settings SET notificationToneUri =:notificationToneUri, isVibrate =:isVibrate, priority =:priority" )
    void updateNotificationSettings(Uri notificationToneUri, boolean isVibrate, int priority);

    @Query("UPDATE app_settings SET isShowHijriDateWidget =:isShowHijriDateWidget, isShowNextReminderWidget =:isDisplayNextReminder" )
    void updateWidgetSettings(boolean isShowHijriDateWidget, boolean isDisplayNextReminder);

}

File: common/src/main/java/com/thesunnahrevival/common/data/local/ReminderDao.kt
package com.thesunnahrevival.common.data.local

import androidx.paging.PagingSource
import androidx.room.*
import com.thesunnahrevival.common.data.model.AppSettings
import com.thesunnahrevival.common.data.model.Reminder
import com.thesunnahrevival.common.data.model.ReminderDate
import kotlinx.coroutines.flow.Flow

@Dao
interface ReminderDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertReminder(reminder: Reminder): Long

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    fun insertRemindersList(remindersList: List<Reminder>)

    @Query("UPDATE reminders_table SET reminderName =:name, reminderInfo =:info, category =:category WHERE id =:id")
    suspend fun updateReminder(id: Int, name: String, info: String, category: String)

    @Delete
    suspend fun deleteReminder(reminder: Reminder)

    @Query(
        "SELECT EXISTS (SELECT * FROM reminders_table WHERE (" +
                "category != :excludeCategory " +
                "AND ((day == :day AND month == :month AND year == :year) OR (day == :day AND month == 12 AND year == 0) OR day == 0 OR customScheduleDays LIKE '%' || :numberOfTheWeekDay || '%')) LIMIT 1)"
    )
    fun thereRemindersOnDay(
        excludeCategory: String,
        numberOfTheWeekDay: String,
        day: Int,
        month: Int,
        year: Int
    ): Boolean

    @Query("SELECT * FROM reminders_table WHERE ((day == :day AND month == :month AND year == :year) OR (day == :day AND month == 12 AND year == 0) OR day == 0 OR customScheduleDays LIKE '%' || :numberOfTheWeekDay || '%') AND (category LIKE '%' || :category || '%') ORDER BY isComplete, timeInSeconds")
    fun getRemindersOnDay(
        numberOfTheWeekDay: String,
        day: Int,
        month: Int,
        year: Int,
        category: String
    ): PagingSource<Int, Reminder>

    @Query("SELECT * FROM reminders_table WHERE ((day == :day AND month == :month AND year == :year) OR (day == :day AND month == 12 AND year == 0) OR day == 0 OR customScheduleDays LIKE '%' || :numberOfTheWeekDay || '%') AND isEnabled ORDER BY timeInSeconds")
    fun getRemindersOnDayValue(
        numberOfTheWeekDay: String,
        day: Int,
        month: Int,
        year: Int
    ): List<Reminder>

    @Query("SELECT (timeInSeconds + (offsetInMinutes * 60)) AS time FROM reminders_table WHERE time > :offsetFromMidnight AND ((day == :day AND month == :month AND year == :year) OR (day == :day AND month == 12 AND year == 0) OR day == 0 OR customScheduleDays LIKE '%' || :numberOfTheWeekDay || '%') AND isEnabled ORDER BY time")
    suspend fun getNextTimeForReminderForDay(
        offsetFromMidnight: Long,
        numberOfTheWeekDay: String,
        day: Int,
        month: Int,
        year: Int
    ): Long?


    @Query("SELECT * FROM reminders_table WHERE (timeInSeconds + (offsetInMinutes * 60)) == :timeForReminder AND ((day == :day AND month == :month AND year == :year) OR (day == :day AND month == 12 AND year == 0) OR day == 0 OR customScheduleDays LIKE '%' || :numberOfTheWeekDay || '%') AND isEnabled ORDER BY (timeInSeconds + (offsetInMinutes * 60))")
    suspend fun getNextScheduledRemindersForDay(
        timeForReminder: Long,
        numberOfTheWeekDay: String,
        day: Int,
        month: Int,
        year: Int
    ): List<Reminder>

    @Query(
        "SELECT EXISTS " +
                "(SELECT * FROM reminders_table WHERE category == :prayerCategory AND " +
                "id LIKE '-' || :id || '%')"
    )
    fun therePrayerRemindersOnDay(prayerCategory: String, id: String): Boolean

    @Query("SELECT * FROM reminders_table WHERE (category ==:categoryName AND (day == :day AND month == :month AND year =:year)) ORDER BY timeInSeconds")
    fun getPrayerTimesValue(day: Int, month: Int, year: Int, categoryName: String): List<Reminder>

    @Query(
        "SELECT day, month, year FROM reminders_table WHERE " +
                "((day >= :day AND month == :month AND year == :year) OR " +
                "(day >= 1 AND month > :month AND year == :year) OR " +
                "(day >= 1 AND month >= 1 AND year > :year)) " +
                "AND id <= -1019700 "
    )
    fun getUpcomingPrayerDates(day: Int, month: Int, year: Int): List<ReminderDate>

    @Query(
        "UPDATE reminders_table SET offsetInMinutes =:offsetValue, reminderInfo =:reminderInfo," +
                " isEnabled = :isEnabled, isComplete = :isComplete" +
                " WHERE id == :reminderId"
    )
    suspend fun updatePrayerTimeDetails(
        reminderInfo: String?,
        offsetValue: Int,
        isEnabled: Boolean,
        isComplete: Boolean,
        reminderId: Int
    )

    @Query(
        "UPDATE reminders_table SET reminderName = :newPrayerName WHERE " +
                "reminderName == :oldPrayerName"
    )
    suspend fun updatePrayerNames(oldPrayerName: String, newPrayerName: String)

    @Query(
        "UPDATE reminders_table SET timeInSeconds = :timeInSeconds, isEnabled = :isEnabled," +
                " offsetInMinutes = :offsetInMinutes WHERE id == :id"
    )
    suspend fun updateGeneratedPrayerTime(
        id: Int,
        timeInSeconds: Long,
        isEnabled: Boolean,
        offsetInMinutes: Int
    )

    @Query("DELETE FROM reminders_table WHERE id < -1019700")
    suspend fun deleteAllPrayerTimes()

    @Query("UPDATE reminders_table SET category =:newCategory WHERE category == :deletedCategory")
    suspend fun updateCategory(deletedCategory: String, newCategory: String)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertSettings(settings: AppSettings)

    @Query("SELECT * FROM app_settings")
    fun getAppSettings(): Flow<AppSettings?>

    @Query("SELECT * FROM app_settings")
    fun getAppSettingsValue(): AppSettings?

    @Update
    suspend fun updateAppSettings(appSettings: AppSettings)

    @Query("UPDATE app_settings SET isShowHijriDateWidget =:isShowHijriDateWidget, isShowNextReminderWidget =:isDisplayNextReminder")
    suspend fun updateWidgetSettings(isShowHijriDateWidget: Boolean, isDisplayNextReminder: Boolean)
}